### ISSUE TITLE
Spring Boot 1.4.0 EndpointMBeanExportAutoConfiguration: NoUniqueBeanDefinitionException: No qualifying bean of type [com.fasterxml.jackson.databind.ObjectMapper] is defined: expected single matching bean but found 3: _halObjectMapper,objectMapper,halObjectMapper

### ISSUE THREAD

**Issue by ceefour** [spring-projects/spring-boot#6517]
This did not happen with Spring Boot 1.3.6.

```
2016-07-31 02:17:04.351  WARN 4744 --- [           main] ationConfigEmbeddedWebApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.springframework.boot.actuate.autoconfigure.EndpointMBeanExportAutoConfiguration': Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.boot.actuate.autoconfigure.EndpointMBeanExportAutoConfiguration$$EnhancerBySpringCGLIB$$58b7686c]: Constructor threw exception; nested exception is org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type [com.fasterxml.jackson.databind.ObjectMapper] is defined: expected single matching bean but found 3: _halObjectMapper,objectMapper,halObjectMapper
2016-07-31 02:17:04.351  INFO 4744 --- [           main] o.s.j.e.a.AnnotationMBeanExporter        : Unregistering JMX-exposed beans on shutdown
2016-07-31 02:17:04.390  INFO 4744 --- [           main] o.a.camel.spring.SpringCamelContext      : Apache Camel 2.17.2 (CamelContext: camel-1) is shutting down
2016-07-31 02:17:04.398  INFO 4744 --- [           main] o.a.camel.spring.SpringCamelContext      : Apache Camel 2.17.2 (CamelContext: camel-1) uptime 
2016-07-31 02:17:04.401  INFO 4744 --- [           main] o.a.camel.spring.SpringCamelContext      : Apache Camel 2.17.2 (CamelContext: camel-1) is shutdown in 0.009 seconds
2016-07-31 02:17:04.406 DEBUG 4744 --- [           main] MongoPersonCustomerRoleHistoryRepository : Shutting down 1 MongoPersonCustomerRoleHistoryRepository beans: [satukancinta]
2016-07-31 02:17:04.412  INFO 4744 --- [           main] kancinta_dev/personCustomerRoleHistory/2 : Shutting down personCustomerRoleHistory MongoDB repository
2016-07-31 02:17:04.412  INFO 4744 --- [           main] MongoPersonCustomerRoleHistoryRepository : Shutted down 1 MongoPersonCustomerRoleHistoryRepository tenant beans: [satukancinta]
2016-07-31 02:17:04.415  INFO 4744 --- [           main] atukancinta_satukancinta_dev/personImage : Shutting down MongoImageRepository person
2016-07-31 02:17:04.417 DEBUG 4744 --- [           main] .s.c.t.TenantBeans/MongoPersonRepository : Shutting down 1 MongoPersonRepository beans: [satukancinta]
2016-07-31 02:17:04.417  INFO 4744 --- [           main] y/satukancinta_satukancinta_dev/person/2 : Shutting down person MongoDB repository
2016-07-31 02:17:04.417  INFO 4744 --- [           main] .s.c.t.TenantBeans/MongoPersonRepository : Shutted down 1 MongoPersonRepository tenant beans: [satukancinta]
2016-07-31 02:17:04.419  INFO 4744 --- [           main] o.s.s.quartz.SchedulerFactoryBean        : Shutting down Quartz Scheduler
2016-07-31 02:17:04.419  INFO 4744 --- [           main] org.quartz.core.QuartzScheduler          : Scheduler scheduler_$_NON_CLUSTERED shutting down.
2016-07-31 02:17:04.419  INFO 4744 --- [           main] org.quartz.core.QuartzScheduler          : Scheduler scheduler_$_NON_CLUSTERED paused.
2016-07-31 02:17:04.736  INFO 4744 --- [           main] org.quartz.core.QuartzScheduler          : Scheduler scheduler_$_NON_CLUSTERED shutdown complete.
2016-07-31 02:17:04.736  INFO 4744 --- [           main] tukancinta_satukancinta_dev/articleImage : Shutting down MongoImageRepository article
2016-07-31 02:17:04.743  INFO 4744 --- [           main] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2016-07-31 02:17:04.771  INFO 4744 --- [           main] utoConfigurationReportLoggingInitializer : 

Error starting ApplicationContext. To display the auto-configuration report enable debug logging (start with --debug)


2016-07-31 02:17:04.812 ERROR 4744 --- [           main] o.s.b.d.LoggingFailureAnalysisReporter   : 

***************************
APPLICATION FAILED TO START
***************************

Description:

Constructor in org.springframework.boot.actuate.autoconfigure.EndpointMBeanExportAutoConfiguration required a single bean, but 3 were found:
	- _halObjectMapper: defined in null
	- objectMapper: defined by method 'objectMapper' in class path resource [org/springframework/boot/autoconfigure/data/rest/SpringBootRepositoryRestMvcConfiguration.class]
	- halObjectMapper: defined by method 'halObjectMapper' in class path resource [org/springframework/boot/autoconfigure/data/rest/SpringBootRepositoryRestMvcConfiguration.class]


Action:

Consider marking one of the beans as @Primary, updating the consumer to accept multiple beans, or using @Qualifier to identify the bean that should be consumed


Process finished with exit code 1
```

Related to #5984, but in this case there's no custom `ObjectMapper` at all.

`_halObjectMapper` is also weird, why "defined in null"?

**Comment by snicoll** [comment]
Duplicates #5984

**Comment by ceefour** [comment]
@snicoll I wonder how this is a duplicate? I mentioned above that it's related to #5984, but there is no custom `ObjectMapper` in this case, so I don't know where to put the `@Primary`...

**Comment by snicoll** [comment]
Share a sample where you don't actually inject the `ObjectMapper` anywhere and I am happy to have a look to it.

**Comment by bitsofinfo** [comment]
Spring data rest guys say the issue lies w/ spring-boot

https://github.com/spring-projects/spring-boot/issues/6529
https://jira.spring.io/browse/DATAREST-865

**Comment by manuel-sousa** [comment]
I've come up with the same issue and I've been reading related issues on both spring boot and spring data rest where all of them are closed and say the other party needs to fix.

Could someone please point me to the fix? Since i'm not using JMX the workaround works for me but it's a workaround and the latent issue seems to be there.

**Comment by wilkinsona** [comment]
@manuel-sousa I don't think this issue has anything to do with JMX. Are you sure you're seeing the same problem? If you are, you either need to mark your `ObjectMapper` bean as `@Primary`, or you need to use a `RepositoryRestConfigurer` rather than subclassing Data REST's configuration class. If neither of these apply, please open a new issue with a sample that reproduces your problem.

**Comment by manuel-sousa** [comment]
@wilkinsona thanks for the prompt reply. The JMX exclusion was the workaround (`@SpringBootApplication(exclude = EndpointMBeanExportAutoConfiguration.class)`) that also fixed the issue in my case.

In my code don't declare any `ObjectMapper`, similar to original request and i get the exact same error. I also don't subclass the `RepositoryRestConfigurer`.

I've setup the simplest project i could that would reproduce the error and made it available at https://github.com/manuel-sousa/springboot-1.4-error

Doing so it seems to be a conjunction of factors if i have `@EnableHypermediaSupport(type = EnableHypermediaSupport.HypermediaType.HAL)` and the `spring-boot-starter-actuator` dependency the error occurs when extending `RepositoryRestConfigurerAdapter`.

On the git I've also left (commented) one autowire issue that also appeared when upgrading to 1.4, this one doesn't need neither the dependency nor the `@EnableHypermediaSupport`.

**Comment by wilkinsona** [comment]
@manuel-sousa Thanks.

> In my code don't declare any `ObjectMapper`

Strictly speaking, you do. `@EnableHypermediaSupport` causes Spring HATEOAS to create an `ObjectMapper` bean for you. And, because you've referenced it in your configuration, it's processed before any auto-configuration. This causes Boot's Jackson auto-configuration to back off leaving the context without a `@Primary` `ObjectMapper`.

You don't need to use `@EnableHypermediaSupport` as Boot will auto-configure it for you. If you use the auto-configuration, Boot will ensure that it's applied after Jackson auto-configuration so the `@Primary` `ObjectMapper` is still created.

**Comment by manuel-sousa** [comment]
@wilkinsona thanks, will test the application without it as I don't recall the reason I added it before. Might be some leftover from some issue. I recall the `Validator` was and doesn't seem to be needed now either.

As a side note, this and validator were working before I migrated from 1.3.6 to 1.4 though and this one only happens with the dependency to `spring-boot-starter-actuator` present.  Feels odd to get an error in data-rest, when using hateoas and having a dependency for easier management  ;)

ps - thanks for all your work with springboot
